pkg/testcase/upgrademanual.go

```diff
@@ -22,66 +28,92 @@ func TestUpgradeClusterManual(cluster *shared.Cluster, k8sClient *k8s.Client, ve
-	var wg sync.WaitGroup
-
-	var ips []string
-	if nodeType == "server" {
-		ips = cluster.ServerIPs
-	} else {
-		ips = cluster.AgentIPs
+func upgradeProduct(product, nodeType, installType, ip string) error {
+	upgradeCommand := shared.GetInstallCmd(product, installType, nodeType)
+	shared.LogLevel("info", "Upgrading %s %s: %s", ip, nodeType, upgradeCommand)
+	if _, err := shared.RunCommandOnNode(upgradeCommand, ip); err != nil {
+		shared.LogLevel("error", "error running cmd on %s %s: %v", nodeType, ip, err)
+		return err
+	}
+	status := []shared.ServiceAction{
+		{
+			Service:  product,
+			Action:   "status",
+			NodeType: nodeType,
+		},
 	}
 
-	errCh := make(chan error, len(ips))
-
-	upgradeCommand := shared.GetInstallCmd(cluster.Config.Product, installType, nodeType)
-
-	for _, ip := range ips {
-		wg.Add(1)
-		go func(ip, upgradeCommand string) {
-			defer wg.Done()
-
-			shared.LogLevel("info", "Upgrading %s %s: %s", ip, nodeType, upgradeCommand)
+	if product == "rke2" {
+		shared.LogLevel("info", "Waiting for 2 mins after installing upgrade...")
+		time.Sleep(2 * time.Minute)
+		ms := shared.NewManageService(3, 3)
+		restart := []shared.ServiceAction{
+			{
+				Service:  product,
+				Action:   "restart",
+				NodeType: nodeType,
+			},
+		}
+		output, err := ms.ManageService(ip, restart)
+		if output != "" {
+			Expect(output).To(ContainSubstring("active "),
+				fmt.Sprintf("error starting %s service for %s node ip: %s", product, nodeType, ip))
+		}
+		Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("error starting %s service on %s", product, ip))
 
-			if _, err := shared.RunCommandOnNode(upgradeCommand, ip); err != nil {
-				shared.LogLevel("warn", "upgrading %s %s: %v", nodeType, ip, err)
-				errCh <- err
-				return
-			}
+		shared.LogLevel("info", "Waiting for 3 mins after restarting service...")
+		time.Sleep(3 * time.Minute)
 
-			shared.LogLevel("info", "Restarting %s: %s", nodeType, ip)
-			err := shared.RestartCluster(cluster.Config.Product, ip)
-			if err != nil {
-				return
-			}
-		}(ip, upgradeCommand)
+		output, err = ms.ManageService(ip, status)
+		if output != "" {
+			Expect(output).To(ContainSubstring("active "),
+				fmt.Sprintf("error checking status %s service for %s node ip: %s", product, nodeType, ip))
+		}
+		Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("error checking status %s service on %s", product, ip))
 	}
 
-	wg.Wait()
-	close(errCh)
-
-	ok, err := k8sClient.CheckClusterHealth(0)
-	if err != nil {
-		return fmt.Errorf("error waiting for cluster to be ready: %w", err)
-	}
-	if !ok {
-		return errors.New("cluster is not healthy")
+	if product == "k3s" {
+		shared.LogLevel("info", "Waiting for 1 mins after installing upgrade...")
+		time.Sleep(1 * time.Minute)
+		ms := shared.NewManageService(3, 1)
+		output, err := ms.ManageService(ip, status)
+		if output != "" {
+			Expect(output).To(ContainSubstring("active "),
+				fmt.Sprintf("error starting %s service for %s node ip: %s", product, nodeType, ip))
```

Comment:
```
shared.LogLevel("error", "") // ?
```

Reply:
Added expect here so no need to log, will update the expect message